test/test_utils.py

```diff
@@ -0,0 +1,321 @@
+        file_to_dump_stdout: Optional[io.FileIO] = None,
+        substitute_stdout_lines: Optional[List[Tuple[bytes, bytes]]] = None,
+        stderr_to_stdout: bool = False,
+        env: Optional[dict] = None,
+        stdout_holder: Optional[List[bytes]] = None) -> subprocess.Popen:
+    if substitute_stdout_lines is None:
+        substitute_stdout_lines = []
+    if env is None:
+        env = os.environ
+    print_debug('Running:', ' '.join(args))
+    iterate_stdout = False
+    kwargs = {}
+    if hide_output:
+        kwargs['stdout'] = subprocess.DEVNULL
+    if hide_stderr:
+        kwargs['stderr'] = subprocess.DEVNULL
+    if file_to_dump_stdout is not None or len(substitute_stdout_lines) > 0 \
+            or stdout_holder is not None:
+        iterate_stdout = True
+        kwargs['stdout'] = subprocess.PIPE
+    if stderr_to_stdout:
+        kwargs['stderr'] = subprocess.STDOUT
+
+    subp = subprocess.Popen(args, bufsize=1, env=env, **kwargs)
+    # Reopen stdout as to be able to output bytes (to avoid conversion to
+    # strings. as we don't know the encoding of whatever we're running)
+    should_reopen_stdout = iterate_stdout and not hide_output
+    line_so_far = []
+    if iterate_stdout:
+        with _maybe_reopen_stdout(
+               should_reopen_stdout=should_reopen_stdout) as stdo:
+
+            def write_to_stdo(bs: bytes):
+                if not hide_output:
+                    stdo.write(bs)
+                if file_to_dump_stdout is not None:
+                    file_to_dump_stdout.write(bs)
+                if stdout_holder is not None:
+                    stdout_holder.append(bs)
+
+            c: bytes = subp.stdout.read(1)
+            while len(c) > 0:
+                if len(substitute_stdout_lines) == 0:
+                    write_to_stdo(c)
+                else:
+                    line_so_far.append(c)
+                    if c[0] == bytes(os.linesep, 'utf-8')[0]:
+                        # Pycharm has the wrong return types for Popen
+                        # noinspection PyTypeChecker
+                        line = b''.join(line_so_far)
+                        line_so_far = []
+                        for m, r in substitute_stdout_lines:
+                            line = re.sub(m, r, line)
+                        write_to_stdo(line)
+                c = subp.stdout.read(1)
+
+            if len(substitute_stdout_lines) > 0 and len(line_so_far) > 0:
+                line = b''.join(line_so_far)
+                for m, r in substitute_stdout_lines:
+                    line = re.sub(m, r, line)
+                write_to_stdo(line)
+    subp.wait()
+    print_debug('Return code is:', subp.returncode)
+    if file_to_dump_stdout is not None:
+        file_to_dump_stdout.flush()
+    if fail_on_failure:
+        if subp.returncode != 0:
+            traceback.print_stack()
+            print_error('Error executing:', ' '.join(args))
+            sys.exit(subp.returncode)
+    return subp
```

Comment:
`execute_command` everywhere is a smell. I'd like to limit it to the things we just can't do with Python, and isolate it to those few areas. Docker stuff can be done with docker-py, and file manipulation stuff with `shutil`.

Executing `git diff` probably is better done with the shell than a library, but even so, it could go into *git.py* and just use `subprocess.run`.